ENT-11065: Added ignoreInaccessibleClasses flag to ReplaceableObjectKryo

shamsasari opened the pull request:
This ignores classes which are inaccessible to the calling module for reflection and so doesn't use their `writeReplace`/`readReplace` methods for serialisation. The default is false to preserve backwards compatibility.

This removes the need to add `add-opens` declarations for the affected packages, especially if the class is in the JDK. Any classes which are ignored by this flag will need to have their own custom serializers.

rick-r3 commented on quasar-core/src/main/java/co/paralleluniverse/io/serialization/kryo/ReplaceableObjectKryo.java:
```diff
@@ -54,17 +56,20 @@ public class ReplaceableObjectKryo extends Kryo {
+    private static final String WRITE_REPLACE = "writeReplace";
+    private static final String READ_RESOLVE = "readResolve";
+
+    private final ClassValue<SerializationMethods> replaceMethodsCache = new ClassValue<>() {
```
I think there is potential bloat from having this be non-static.  We create quite a few instances (I think we use a pool, so likely one per thread).

This is so the value of `ignoreInaccessibleClasses` is correctly taken into account for cached items?  We could just have two (still static) caches, and get from the correct one dependent on value of `ignoreInaccessibleClasses`?

rick-r3 reviewed (commented):
Could you talk a bit about how this helps?  I understand what it does differently, but when do you intend to use it / why / why is okay etc ?

shamsasari committed:
ENT-11065: Added ignoreInaccessibleClasses flag to ReplaceableObjectKryo

This ignores classes which are inaccessible to the calling module for reflection and so doesn't use their writeReplace/readReplace methods for serialisation. The default is false to preserve backwards compatibility.

This removes the need to add add-opens declarations for the affected packages, especially if the class is in the JDK. Any classes which are ignored by this flag will need to have their own custom serializers.

shamsasari commented:
PTAL. Let me know if you need me to elaborate more.

shamsasari commented:
Need to reassess my whole strategy. If I go with a more lazy initialisation of Kryo in the node then this change may not be needed.

shamsasari commented:
Closing for now.
